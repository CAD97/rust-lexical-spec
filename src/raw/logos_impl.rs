use logos::Logos;

#[allow(nonstandard_style)]
#[derive(Logos, Debug, Eq, PartialEq)]
enum TokenKind {
    #[regex(r"//[^\n]*")]
    line_comment,
    #[regex(r"\p{Pattern_White_Space}+")]
    whitespace,
    #[regex(r"[_\p{XID_Start}]\p{XID_Continue}*")]
    identifier,
    #[regex(r"r#[_\p{XID_Start}]\p{XID_Continue}*")]
    raw_identifier,
    #[regex(r"[\p{XID_Continue}--_0-9\p{XID_Start}]\p{XID_Continue}*")]
    identifier_fragment,
    #[regex(r"'[_\p{XID_Start}]\p{XID_Continue}*")]
    lifetime,
    #[regex(r"0b[_0-9]*")]
    binary_integer,
    #[regex(r"0b[_0-9]*[eE][+-]?[_0-9]*")]
    binary_float,
    #[regex(r"0o[_0-9]*")]
    octal_integer,
    #[regex(r"0o[_0-9]*[eE][+-]?[_0-9]*")]
    octal_float,
    #[regex(r"0x[_0-9a-fA-F]*")]
    hexadecimal_integer,
    #[regex(r"[0-9][_0-9]*")]
    decimal_integer,
    #[regex(r"[0-9][_0-9]*[eE][+-]?[_0-9]*")]
    decimal_float,
    #[regex(r#"'(?:[^\t\n\r\\']|\\['"]|\\[nrt\\0]|\\x[0-7][0-9a-fA-F]|\\u\{(?:10|[0-9])[0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?\})'"#)]
    character,
    #[regex(r#"'(?:[[:ascii:]--\t\n\r\\']|\\['"]|\\[nrt\\0]|\\x[0-9a-fA-F][0-9a-fA-F])'"#)]
    byte,
    #[regex(r#""(?:[^\t\n\r\\"]|\r\n|\\\n\p{Pattern_White_Space}*|\\['"]|\\[nrt\\0]|\\x[0-7][0-9a-fA-F]|\\u\{(?:10|[0-9])[0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?[0-9a-fA-F]?\})*""#)]
    string,
    #[regex(r#"b"(?:[[:ascii:]--\r\\"]|\r\n|\\\n\p{Pattern_White_Space}*|\\['"]|\\[nrt\\0]|\\x[0-9a-fA-F][0-9a-fA-F])*""#)]
    byte_string,
    #[regex(r"!")]
    exclamation,
    #[regex(r"#")]
    pound,
    #[regex(r"\$")]
    dollar,
    #[regex(r"%")]
    percent,
    #[regex(r"&")]
    ampersand,
    #[regex(r"\(")]
    open_parenthesis,
    #[regex(r"\)")]
    close_parenthesis,
    #[regex(r"\*")]
    star,
    #[regex(r"\+")]
    plus,
    #[regex(r",")]
    comma,
    #[regex(r"-")]
    minus,
    #[regex(r"\.")]
    dot,
    #[regex(r"/")]
    slash,
    #[regex(r":")]
    colon,
    #[regex(r";")]
    semicolon,
    #[regex(r"<")]
    less,
    #[regex(r"=")]
    equal,
    #[regex(r">")]
    greater,
    #[regex(r"\?")]
    question,
    #[regex(r"@")]
    at,
    #[regex(r"\[")]
    open_bracket,
    #[regex(r"\]")]
    close_bracket,
    #[regex(r"\^")]
    circumflex,
    #[regex(r"\{")]
    open_brace,
    #[regex(r"\|")]
    bar,
    #[regex(r"\}")]
    close_brace,
    #[regex(r"~")]
    tilde,
    #[regex(r"/\*", parse_block_comment)]
    block_comment,
    #[regex(r#"r[#"]"#, parse_raw_string)]
    raw_string,
    #[regex(r#"br[#"]"#, parse_raw_byte_string)]
    raw_byte_string,
    #[error]
    Error,
}

pub struct Lexer;

impl super::Lexer for Lexer {
    fn lex(s: &str) -> Result<super::Token, ()> {
        let lexer = TokenKind::lexer(s).spanned();
        let (kind, span) = { lexer }.next().ok_or(())?;
        assert_eq!(span.start, 0);
        super::Class::ALL
            .get(kind as usize)
            .ok_or(())
            .map(|&class| super::Token {
                class,
                length: span.end,
            })
    }
}

fn parse_block_comment(lex: &mut logos::Lexer<TokenKind>) -> Result<(), &'static str> {
    let mut depth: usize = 1;
    while depth > 0 {
        let interesting_at = lex
            .remainder()
            .find(&['/', '*'][..])
            .ok_or("exhausted source in block comment")?;
        lex.bump(interesting_at);
        if lex.remainder().starts_with("/*") {
            lex.bump(2);
            depth += 1;
        } else if lex.remainder().starts_with("*/") {
            lex.bump(2);
            depth -= 1;
        } else {
            lex.bump(1);
        }
    }
    Ok(())
}

fn parse_raw_string(lex: &mut logos::Lexer<TokenKind>) -> Result<(), &'static str> {
    let hash_count: usize = match lex.slice() {
        r#"r""# => 0,
        r#"r#"# => {
            let mut hashes = 1;
            loop {
                match lex.remainder().chars().next() {
                    Some('#') => {
                        lex.bump(1);
                        hashes += 1;
                    }
                    Some('"') => {
                        lex.bump(1);
                        break;
                    }
                    Some(_) => Err("invalid char in raw string opening fence")?,
                    None => Err("exhausted source in raw string")?,
                }
            }
            hashes
        }
        _ => Err("unreachable")?,
    };

    loop {
        match lex.remainder().chars().next() {
            Some('"') => {
                lex.bump(1);
                let mut hashes_seen: usize = 0;
                loop {
                    if hashes_seen == hash_count {
                        return Ok(());
                    }
                    match lex.remainder().chars().next() {
                        Some('#') => {
                            lex.bump(1);
                            hashes_seen += 1;
                        }
                        _ => break,
                    }
                }
            }
            Some('\r') => {
                lex.bump(1);
                match lex.remainder().chars().next() {
                    Some('\n') => lex.bump(1),
                    _ => Err("bare CR not allowed in raw string")?,
                }
            }
            Some(c) => lex.bump(c.len_utf8()),
            None => Err("exhausted source in raw string")?,
        }
    }
}

fn parse_raw_byte_string(lex: &mut logos::Lexer<TokenKind>) -> Result<(), &'static str> {
    let hash_count: usize = match lex.slice() {
        r#"br""# => 0,
        r#"br#"# => {
            let mut hashes = 1;
            loop {
                match lex.remainder().chars().next() {
                    Some('#') => {
                        lex.bump(1);
                        hashes += 1;
                    }
                    Some('"') => {
                        lex.bump(1);
                        break;
                    }
                    Some(_) => Err("invalid char in raw string opening fence")?,
                    None => Err("exhausted source in raw string")?,
                }
            }
            hashes
        }
        _ => Err("unreachable")?,
    };

    loop {
        match lex.remainder().chars().next() {
            Some('"') => {
                lex.bump(1);
                let mut hashes_seen: usize = 0;
                loop {
                    if hashes_seen == hash_count {
                        return Ok(());
                    }
                    match lex.remainder().chars().next() {
                        Some('#') => {
                            lex.bump(1);
                            hashes_seen += 1;
                        }
                        _ => break,
                    }
                }
            }
            Some('\r') => {
                lex.bump(1);
                match lex.remainder().chars().next() {
                    Some('\n') => lex.bump(1),
                    _ => Err("bare CR not allowed in raw string")?,
                }
            }
            Some(c) => lex.bump(c.len_utf8()),
            None => Err("exhausted source in raw string")?,
        }
    }
}
