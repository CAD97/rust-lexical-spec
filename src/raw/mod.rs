#[cfg(feature = "logos")]
pub mod logos_impl;
pub mod reference_impl;

#[allow(nonstandard_style)]
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Class {
    line_comment,
    whitespace,
    identifier,
    raw_identifier,
    identifier_fragment,
    lifetime,
    binary_integer,
    binary_float,
    octal_integer,
    octal_float,
    hexadecimal_integer,
    decimal_integer,
    decimal_float,
    character,
    byte,
    string,
    byte_string,
    exclamation,
    pound,
    dollar,
    percent,
    ampersand,
    open_parenthesis,
    close_parenthesis,
    star,
    plus,
    comma,
    minus,
    dot,
    slash,
    colon,
    semicolon,
    less,
    equal,
    greater,
    question,
    at,
    open_bracket,
    close_bracket,
    circumflex,
    open_brace,
    bar,
    close_brace,
    tilde,
    block_comment,
    raw_string,
    raw_byte_string,
}
pub use Class::*;

impl Class {
    pub const COUNT: usize = raw_byte_string as usize + 1;
    pub const ALL: [Class; Class::COUNT] = [
        line_comment,
        whitespace,
        identifier,
        raw_identifier,
        identifier_fragment,
        lifetime,
        binary_integer,
        binary_float,
        octal_integer,
        octal_float,
        hexadecimal_integer,
        decimal_integer,
        decimal_float,
        character,
        byte,
        string,
        byte_string,
        exclamation,
        pound,
        dollar,
        percent,
        ampersand,
        open_parenthesis,
        close_parenthesis,
        star,
        plus,
        comma,
        minus,
        dot,
        slash,
        colon,
        semicolon,
        less,
        equal,
        greater,
        question,
        at,
        open_bracket,
        close_bracket,
        circumflex,
        open_brace,
        bar,
        close_brace,
        tilde,
        block_comment,
        raw_string,
        raw_byte_string,
    ];
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Token {
    pub class: Class,
    pub length: usize,
}

pub trait Lexer {
    fn lex(s: &str) -> Result<Token, ()>;
    fn lex_all(s: &str) -> Result<Vec<Token>, ()> {
        let mut s = s;
        std::iter::from_fn(|| {
            if s.is_empty() {
                return None;
            } else {
                match Self::lex(s) {
                    Ok(token) => {
                        s = &s[token.length..];
                        Some(Ok(token))
                    }
                    Err(()) => {
                        s = "";
                        Some(Err(()))
                    }
                }
            }
        })
        .collect()
    }
}
