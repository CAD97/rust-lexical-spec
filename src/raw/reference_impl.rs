use {
    super::*,
    once_cell::sync::Lazy,
    regex::{Regex, RegexSet},
    std::panic::catch_unwind,
};

static REGEX_SET: Lazy<RegexSet> = Lazy::new(|| {
    RegexSet::new(vec![
        r"\A//[^\n]*",                                               // line_comment
        r"\A\p{Pattern_White_Space}+",                               // whitespace
        r"\A[_\p{XID_Start}]\p{XID_Continue}*",                      // identifier
        r"\Ar#[_\p{XID_Start}]\p{XID_Continue}*",                    // raw_identifier
        r"\A[\p{XID_Continue}--_0-9\p{XID_Start}]\p{XID_Continue}*", // identifier_fragment
        r"\A'[_\p{XID_Start}]\p{XID_Continue}*",                     // lifetime
        r"\A0b[_0-9]*",                                              // binary_integer
        r"\A0b[_0-9]*[eE][+-]?[_0-9]*",                              // binary_float
        r"\A0o[_0-9]*",                                              // octal_integer
        r"\A0o[_0-9]*[eE][+-]?[_0-9]*",                              // octal_float
        r"\A0x[_0-9a-fA-F]*",                                        // hexadecimal_integer
        r"\A0x[_0-9a-fA-F]*[eE][+-]?[_0-9a-fA-F]*",                  // hexadecimal_float
        r"\A[0-9][_0-9]*",                                           // decimal_integer
        r"\A[0-9][_0-9]*[eE][+-]?[_0-9]*",                           // decimal_float
        concat!(
            r#"\A'(?:[^\t\n\r\\']"#,
            r#"|\\['"]"#,                              // quote_escape
            r#"|\\[nrt\\0]"#,                          // named_escape
            r#"|\\x[0-7][0-9a-fA-F]"#,                 // ascii_escape
            r#"|\\u\{(?:10|[0-9])[0-9a-fA-F]{0,4}\}"#, // unicode_escape
            r#")'"#
        ), // character
        concat!(
            r#"\A'(?:[[:ascii:]--\t\n\r\\']"#,
            r#"|\\['"]"#,            // quote_escape
            r#"|\\[nrt\\0]"#,        // named_escape
            r#"|\\x[0-9a-fA-F]{2}"#, // byte_escape
            r#")'"#
        ), // byte
        concat!(
            r#"\A"(?:[^\t\n\r\\"]"#,
            r#"|\r\n"#,
            r#"|\\\n\p{Pattern_White_Space}*"#, // indent_escape
            r#"|\\['"]"#,                       // quote_escape
            r#"|\\[nrt\\0]"#,                   // named_escape
            r#"|\\x[0-7][0-9a-fA-F]"#,          // ascii_escape
            r#"|\\u\{(?:10|[0-9])[0-9a-fA-F]{0,4}\}"#, // unicode_escape
            r#")*""#
        ), // string
        concat!(
            r#"\Ab"(?:[[:ascii:]--\r\\"]"#,
            r#"|\r\n"#,
            r#"|\\\n\p{Pattern_White_Space}*"#, // indent_escape
            r#"|\\['"]"#,                       // quote_escape
            r#"|\\[nrt\\0]"#,                   // named_escape
            r#"|\\x[0-9a-fA-F]{2}"#,            // byte_escape
            r#")*""#
        ), // byte_string
        r"\A!",                                                      // exclamation
        r"\A#",                                                      // pound
        r"\A\$",                                                     // dollar
        r"\A%",                                                      // percent
        r"\A&",                                                      // ampersand
        r"\A\(",                                                     // open_parenthesis
        r"\A\)",                                                     // close_parenthesis
        r"\A\*",                                                     // star
        r"\A\+",                                                     // plus
        r"\A,",                                                      // comma
        r"\A-",                                                      // minus
        r"\A\.",                                                     // dot
        r"\A/",                                                      // slash
        r"\A:",                                                      // colon
        r"\A;",                                                      // semicolon
        r"\A<",                                                      // less
        r"\A=",                                                      // equal
        r"\A>",                                                      // greater
        r"\A\?",                                                     // question
        r"\A@",                                                      // at
        r"\A\[",                                                     // open_bracket
        r"\A\]",                                                     // close_bracket
        r"\A\^",                                                     // circumflex
        r"\A\{",                                                     // open_brace
        r"\A\|",                                                     // bar
        r"\A\}",                                                     // close_brace
        r"\A~",                                                      // tilde
        r"\A/\*",                                                    // block_comment
        r#"\Ar[#"]"#,                                                // raw_string
        r#"\Abr[#"]"#,                                               // raw_byte_string
    ])
    .unwrap()
});

static ALL_REGEX: Lazy<Vec<Regex>> = Lazy::new(|| {
    let mut vec = vec![];
    for i in 0..REGEX_SET.patterns().len() {
        vec.push(Regex::new(&REGEX_SET.patterns()[i]).unwrap());
    }
    vec
});

pub struct Lexer;

impl super::Lexer for Lexer {
    fn lex(s: &str) -> Result<Token, ()> {
        let matches: Vec<Class> = REGEX_SET
            .matches(s)
            .iter()
            .map(|ix| Class::ALL[ix])
            .collect();
        let class = match &*matches {
            [] => Err(())?,
            &[any] => any,
            [line_comment, slash] => line_comment,
            [binary_integer, decimal_integer] => binary_integer,
            [binary_integer, binary_float, decimal_integer] => binary_float,
            [octal_integer, decimal_integer] => octal_integer,
            [octal_integer, octal_float, decimal_integer] => octal_float,
            [hexadecimal_integer, decimal_integer] => hexadecimal_integer,
            [hexadecimal_integer, hexadecimal_float, decimal_integer] => hexadecimal_float,
            [decimal_integer, decimal_float] => decimal_float,
            [lifetime, character] => character,
            [identifier, byte_string] => byte_string,
            [identifier, identifier_fragment] => identifier,
            [slash, block_comment] => block_comment,
            [identifier, raw_string] => raw_string,
            [identifier, raw_byte_string] => raw_byte_string,
            other => unimplemented!("raw lexer matched set {:?}", other),
        };
        Ok(match class {
            block_comment => Token {
                class,
                length: catch_unwind(|| parse_block_comment(s)).map_err(drop)?,
            },
            raw_string => Token {
                class,
                length: catch_unwind(|| parse_raw_string(s)).map_err(drop)?,
            },
            raw_byte_string => Token {
                class,
                length: catch_unwind(|| parse_raw_byte_string(s)).map_err(drop)?,
            },
            _ => {
                let mat = ALL_REGEX[class as usize].find(s).unwrap();
                assert_eq!(mat.start(), 0);
                Token {
                    class,
                    length: mat.end(),
                }
            }
        })
    }
}

// Reminder: these functions are copied from the specification.
// Do not optimize or make idiomatic; their purpose is to be obvious.

pub fn parse_block_comment(s: &str) -> usize {
    let mut chars = s.chars().peekable();
    assert_eq!(chars.next(), Some('/'));
    assert_eq!(chars.next(), Some('*'));

    let mut depth: usize = 1;
    let mut len: usize = 2;

    while depth > 0 {
        match chars.next() {
            Some('/') if matches!(chars.peek(), Some('*')) => {
                chars.next();
                depth += 1;
                len += 2;
            }
            Some('*') if matches!(chars.peek(), Some('/')) => {
                chars.next();
                depth -= 1;
                len += 2;
            }
            Some(c) => len += c.len_utf8(),
            None => panic!("exhausted source in block comment"),
        }
    }

    return len;
}

fn parse_raw_string(s: &str) -> usize {
    let mut chars = s.chars().peekable();
    assert_eq!(chars.next(), Some('r'));

    let mut hashes: usize = 0;
    let mut len: usize = 1;

    loop {
        match chars.next() {
            Some('#') => {
                len += 1;
                hashes += 1;
            }
            Some('"') => {
                len += 1;
                break;
            }
            Some(c) => panic!("invalid char {:?} in raw string opening fence", c),
            None => panic!("exhausted source in raw string"),
        }
    }

    loop {
        match chars.next() {
            Some('"') => {
                len += 1;
                let mut hashes_seen: usize = 0;
                loop {
                    if hashes_seen == hashes {
                        return len;
                    }
                    match chars.peek() {
                        Some('#') => {
                            chars.next();
                            len += 1;
                            hashes_seen += 1;
                        }
                        _ => break,
                    }
                }
            }
            Some('\r') => match chars.next() {
                Some('\n') => len += 2,
                _ => panic!("bare CR not allowed in raw string"),
            },
            Some(c) => len += c.len_utf8(),
            None => panic!("exhausted source in raw string"),
        }
    }
}

fn parse_raw_byte_string(s: &str) -> usize {
    let mut chars = s.chars().peekable();
    assert_eq!(chars.next(), Some('b'));
    assert_eq!(chars.next(), Some('r'));

    let mut hashes: usize = 0;
    let mut len: usize = 2;

    loop {
        match chars.next() {
            Some('#') => {
                len += 1;
                hashes += 1;
            }
            Some('"') => {
                len += 1;
                break;
            }
            Some(c) => panic!("invalid char {:?} in raw byte string opening fence", c),
            None => panic!("exhausted source in raw byte string"),
        }
    }

    loop {
        match chars.next() {
            Some('"') => {
                len += 1;
                let mut hashes_seen: usize = 0;
                loop {
                    if hashes_seen == hashes {
                        return len;
                    }
                    match chars.peek() {
                        Some('#') => {
                            chars.next();
                            len += 1;
                            hashes_seen += 1;
                        }
                        _ => break,
                    }
                }
            }
            Some('\r') => match chars.next() {
                Some('\n') => len += 2,
                _ => panic!("bare CR not allowed in raw byte string"),
            },
            Some(c) if c.is_ascii() => len += 1,
            Some(_) => panic!("raw byte string must be ASCII"),
            None => panic!("exhausted source in raw byte string"),
        }
    }
}
