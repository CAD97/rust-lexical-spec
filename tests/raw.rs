use lexical_specification::raw::{self, reference_impl, Lexer as _};

macro_rules! test_all {
    ($ReferenceLexer:path [$($(#[cfg(feature = $feature:literal)])? $Lexer:path),+ $(,)?] $text:tt) => {
        test_all!(@reference $ReferenceLexer $text);
        $(test_all!(@compare $(#[cfg(feature = $feature)])? $ReferenceLexer, $Lexer $text);)+
    };
    (@reference $Lexer:path [$($text:literal),+ $(,)?]) => {
        $(insta::assert_debug_snapshot!(<$Lexer>::lex_all($text));)+
    };
    (@compare $(#[cfg(feature = $feature:literal)])? $Reference:path, $Lexer:path [$($text:literal),+ $(,)?]) => {
        $(#[cfg(feature = $feature)])?
        {
            $(assert_eq!(<$Reference>::lex_all($text), <$Lexer>::lex_all($text),
                "Reference lexer and {} mismatch on input {:?}",
                stringify!($Lexer), $text);)+
        }
    };
}

#[test]
fn examples_in_informal_reference() {
    test_all! {
        reference_impl::Lexer
        [
            raw::reference_impl::Lexer,
            #[cfg(feature = "logos")] raw::logos_impl::Lexer,
        ]
        [
            // <https://doc.rust-lang.org/stable/reference/keywords.html>
            "as", "break", "const", "continue", "crate", "else", "enum",
            "extern", "false", "fn", "for", "if", "impl", "in", "let",
            "loop", "match", "mod", "move", "mut", "pub", "ref", "return",
            "self", "Self", "static", "struct", "super", "trait", "true",
            "type", "unsafe", "use", "where", "while", "async", "await",
            "dyn", "abstract", "become", "box", "do", "final", "macro",
            "override", "priv", "typeof", "unsized", "virtual", "yield",
            "try", "union", "'static", "dyn",
            // <https://doc.rust-lang.org/stable/reference/comments.html>
            "\
//! A doc comment that applies to the implicit anonymous module of this crate

pub mod outer_module {

//!  - Inner line doc
//!! - Still an inner line doc (but with a bang at the beginning)

/*!  - Inner block doc */
/*!! - Still an inner block doc (but with a bang at the beginning) */

//   - Only a comment
///  - Outer line doc (exactly 3 slashes)
//// - Only a comment

/*   - Only a comment */
/**  - Outer block doc (exactly) 2 asterisks */
/*** - Only a comment */

pub mod inner_module {}

pub mod nested_comments {
    /* In Rust /* we can /* nest comments */ */ */

    // All three types of block comments can contain or be nested inside
    // any other type:

    /*   /* */  /** */  /*! */  */
    /*!  /* */  /** */  /*! */  */
    /**  /* */  /** */  /*! */  */
    pub mod dummy_item {}
}

pub mod degenerate_cases {
    // empty inner line doc
    //!

    // empty inner block doc
    /*!*/

    // empty line comment
    //

    // empty outer line doc
    ///

    // empty block comment
    /**/

    pub mod dummy_item {}

    // empty 2-asterisk block isn't a doc block, it is a block comment
    /***/

}

/* The next one isn't allowed because outer doc comments
    require an item that will receive the doc */

/// Where is my item?
}",
            // <https://doc.rust-lang.org/stable/reference/whitespace.html>
            "\u{0009}", "\u{000A}", "\u{000B}", "\u{000C}", "\u{000D}",
            "\u{0020}", "\u{0085}", "\u{200E}", "\u{200F}", "\u{2028}",
            "\u{2029}",
            // <https://doc.rust-lang.org/stable/reference/tokens.html>
            "98_222", "0xff", "0o77", "0b1111_0000", "123.0E+77", "1_234.0E+18f64", r###""foo""###, r###"r"foo""###, r###""\"foo\"""###, r###"r#""foo""#"###, r###""foo #\"# bar""###, r###"r##"foo #"# bar"##"###, r###""\x52""###, r###""R""###, r###"r"R""###, r###""\\x52""###, r###"r"\x52""###, r###"b"foo""###, r###"br"foo""###, r###"b"\"foo\"""###, r###"br#""foo""#"###, r###"b"foo #\"# bar""###, r###"br##"foo #"# bar"##"###, r###"b"\x52""###, r###"b"R""###, r###"br"R""###, r###"b"\\x52""###, r###"br"\x52""###, "123", "123i32", "123u32", "123_u32", "0xff", "0xff_u8", "0o70", "0o70_i16", "0b1111_1111_1001_0000", "0b1111_1111_1001_0000i64", "0b________1", "0usize", "0invalidSuffix", "123AFB43", "0b0102", "0o0581", "128_i8", "256_u8", "0b_", "0b____", "123.0f64", "0.1f64", "0.1f32", "12E+99_f64", "2.",
        ]
    }
}
