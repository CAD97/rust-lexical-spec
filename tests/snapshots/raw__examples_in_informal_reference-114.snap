---
source: tests/raw.rs
expression: "<reference_impl::Lexer>::lex_all(\"0b____\")"
---
Ok(
    [
        Token {
            class: binary_integer,
            length: 6,
        },
    ],
)
